File: packages/kandi-login/examples/ios/Sources/KandiLoginSDK/AuthManager.swift
// AuthManager.swift
// KandiLoginSDK
//
// Copyright (c) KandiForge. All rights reserved.
// Licensed under the MIT License.

import AuthenticationServices
import Foundation
import Observation

@Observable
@MainActor
public final class AuthManager {

    // MARK: - Observable State

    public private(set) var isAuthenticated: Bool = false
    public private(set) var currentUser: KandiUser?
    public private(set) var isLoading: Bool = false
    public private(set) var errorMessage: String?

    // MARK: - Dependencies

    private let client: KandiLoginClient
    private let tokenStorage: KeychainTokenStorage
    private let config: KandiLoginConfig

    // MARK: - Initialization

    public init(config: KandiLoginConfig) {
        self.config = config
        self.client = KandiLoginClient(authServerUrl: config.authServerUrl)
        self.tokenStorage = KeychainTokenStorage()
    }

    // MARK: - Public Methods

    /// Attempts to restore a previous session by validating the stored access token.
    public func restoreSession() async {
        guard let accessToken = tokenStorage.getAccessToken() else {
            return
        }

        isLoading = true
        defer { isLoading = false }

        do {
            let user = try await client.validateToken(accessToken)
            currentUser = user
            isAuthenticated = true
        } catch {
            // Token may be expired, try refreshing
            await refreshIfNeeded()
        }
    }

    /// Initiates the OAuth login flow for the given provider.
    public func login(
        provider: String,
        presentationContextProvider: ASWebAuthenticationPresentationContextProviding
    ) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let tokenResponse = try await client.login(
                provider: provider,
                callbackScheme: config.callbackScheme,
                presentationContextProvider: presentationContextProvider
            )

            tokenStorage.saveTokens(
                access: tokenResponse.accessToken,
                refresh: tokenResponse.refreshToken
            )

            let user = try await client.validateToken(tokenResponse.accessToken)
            currentUser = user
            isAuthenticated = true
        } catch let error as KandiLoginError {
            if case .authenticationCancelled = error {
                // User cancelled, no error to display
                return
            }
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// Logs in as a test persona (for development use).
    public func loginAsPersona(_ personaId: String) async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let tokenResponse = try await client.loginAs(personaId: personaId)

            tokenStorage.saveTokens(
                access: tokenResponse.accessToken,
                refresh: tokenResponse.refreshToken
            )

            let user = try await client.validateToken(tokenResponse.accessToken)
            currentUser = user
            isAuthenticated = true
        } catch let error as KandiLoginError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    /// Logs out the current user.
    public func logout() async {
        isLoading = true
        defer { isLoading = false }

        if let accessToken = tokenStorage.getAccessToken() {
            try? await client.logout(accessToken: accessToken)
        }

        tokenStorage.clearTokens()
        currentUser = nil
        isAuthenticated = false
        errorMessage = nil
    }

    /// Refreshes the access token if a refresh token is available.
    public func refreshIfNeeded() async {
        guard let refreshToken = tokenStorage.getRefreshToken() else {
            await clearSession()
            return
        }

        do {
            let tokenResponse = try await client.refreshToken(refreshToken)

            tokenStorage.saveTokens(
                access: tokenResponse.accessToken,
                refresh: tokenResponse.refreshToken
            )

            let user = try await client.validateToken(tokenResponse.accessToken)
            currentUser = user
            isAuthenticated = true
        } catch {
            await clearSession()
        }
    }

    /// Seeds test personas on the auth server.
    public func seedTestPersonas() async throws -> SeedResponse {
        return try await client.seedTestPersonas()
    }

    /// Lists available test personas.
    public func listPersonas() async throws -> PersonasResponse {
        return try await client.listPersonas()
    }

    // MARK: - Private Helpers

    private func clearSession() async {
        tokenStorage.clearTokens()
        currentUser = nil
        isAuthenticated = false
    }
}

File: packages/kandi-login/examples/android/app/src/main/java/com/kandiforge/example/sdk/AuthManager.kt
// Copyright (c) KandiForge. MIT License.

package com.kandiforge.example.sdk

import android.content.Context
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class AuthManager(context: Context) {

    private val client = KandiLoginClient()
    private val tokenStorage = SecureTokenStorage(context)

    private val _isAuthenticated = MutableStateFlow(false)
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private val _currentUser = MutableStateFlow<KandiUser?>(null)
    val currentUser: StateFlow<KandiUser?> = _currentUser.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    val loginClient: KandiLoginClient get() = client

    // MARK: - Session Restore

    /**
     * Attempts to restore a previous session by validating the stored access token.
     */
    suspend fun restoreSession() {
        val accessToken = tokenStorage.getAccessToken() ?: return

        _isLoading.value = true
        try {
            val user = client.validateToken(accessToken)
            _currentUser.value = user
            _isAuthenticated.value = true
        } catch (_: Exception) {
            refreshIfNeeded()
        } finally {
            _isLoading.value = false
        }
    }

    // MARK: - Handle OAuth Callback

    /**
     * Processes the callback URI from the OAuth flow (Custom Chrome Tab redirect).
     */
    suspend fun handleCallback(uri: android.net.Uri) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val tokenResponse = client.parseTokensFromCallback(uri)
            tokenStorage.saveTokens(
                access = tokenResponse.accessToken,
                refresh = tokenResponse.refreshToken,
            )
            val user = client.validateToken(tokenResponse.accessToken)
            _currentUser.value = user
            _isAuthenticated.value = true
        } catch (e: KandiLoginException.AuthenticationCancelled) {
            // User cancelled, no error to display
        } catch (e: KandiLoginException) {
            _errorMessage.value = e.message
        } catch (e: Exception) {
            _errorMessage.value = e.message
        } finally {
            _isLoading.value = false
        }
    }

    // MARK: - Test Persona Login

    /**
     * Logs in as a test persona (for development use).
     */
    suspend fun loginAsPersona(personaId: String) {
        _isLoading.value = true
        _errorMessage.value = null
        try {
            val tokenResponse = client.loginAs(personaId)
            tokenStorage.saveTokens(
                access = tokenResponse.accessToken,
                refresh = tokenResponse.refreshToken,
            )
            val user = client.validateToken(tokenResponse.accessToken)
            _currentUser.value = user
            _isAuthenticated.value = true
        } catch (e: KandiLoginException) {
            _errorMessage.value = e.message
        } catch (e: Exception) {
            _errorMessage.value = e.message
        } finally {
            _isLoading.value = false
        }
    }

    // MARK: - Logout

    /**
     * Logs out the current user.
     */
    suspend fun logout() {
        _isLoading.value = true
        try {
            val accessToken = tokenStorage.getAccessToken()
            if (accessToken != null) {
                try {
                    client.logout(accessToken)
                } catch (_: Exception) {
                    // Best-effort server logout
                }
            }
        } finally {
            tokenStorage.clearTokens()
            _currentUser.value = null
            _isAuthenticated.value = false
            _errorMessage.value = null
            _isLoading.value = false
        }
    }

    // MARK: - Refresh

    /**
     * Refreshes the access token if a refresh token is available.
     */
    suspend fun refreshIfNeeded() {
        val refreshToken = tokenStorage.getRefreshToken()
        if (refreshToken == null) {
            clearSession()
            return
        }

        try {
            val tokenResponse = client.refreshToken(refreshToken)
            tokenStorage.saveTokens(
                access = tokenResponse.accessToken,
                refresh = tokenResponse.refreshToken,
            )
            val user = client.validateToken(tokenResponse.accessToken)
            _currentUser.value = user
            _isAuthenticated.value = true
        } catch (_: Exception) {
            clearSession()
        }
    }

    // MARK: - Test Endpoints

    /**
     * Seeds test personas on the auth server.
     */
    suspend fun seedTestPersonas(): SeedResponse {
        return client.seedTestPersonas()
    }

    /**
     * Lists available test personas.
     */
    suspend fun listPersonas(): PersonasResponse {
        return client.listPersonas()
    }

    // MARK: - Private

    private fun clearSession() {
        tokenStorage.clearTokens()
        _currentUser.value = null
        _isAuthenticated.value = false
    }
}
